rfc-174-consolidate-frontend-rendering-apps.md

```diff
@@ -0,0 +1,307 @@
+
+### Translations
+There is existing tech debt for [multiple frontend applications use the same translated phrases]. These are some of the existing issues that will be addressed by having a single rendering app:
+
+1. Different pages on the site can have different translations for the same English word
+2. Some translations will be missing on pages rendered by different frontend applications
+3. Difficult to get visibility on which phrases are translated and which ones aren't
+
+## Proposal
+
+We propose merging the apps into a single rendering app. This will solve some of the issues in this RFC and allow us to solve others in the future - for example, removing `static`.
+
+We propose merging all apps into an existing app. The benefit of this is that even if we aren’t able to consolidate all of the applications, we would still reduce the maintenance burden for each app merged. We also won’t have created anything new, and won’t have to manage any extra conflicting patterns. 
+
+We can steadily move routes, controllers and models from the candidate (the app being merged) into the host (the app we are merging into) until the host is serving all the routes.
+
+We propose using [frontend] as the host app. This is because:
+
+* `frontend` contains some of the more complicated document types like licence_transaction and local_transaction, that have multiple pages and interactive elements, which would be harder to consolidate into another app
+* `frontend` has a one-controller per document type layout, which is preferable for document types with multiple pages as it follows [rails conventions for controllers] to do the orchestration.
+
+[government-frontend] was also considered for the host app, but not chosen because it has a more restrictive presenter pattern that works well for single-page document types. Even though `government-frontend` has many document types and accounts for 50% of the frontend traffic, it would be easier to move the single-page document types from `government-frontend` to `frontend` than it would to try and force the document types in `frontend` into the presenter pattern.
+
+### Testing standardisation
+We propose standardising around RSpec as the test suite for the frontend, ideally [modernising the tests], for example by switching to request tests rather than controller tests and system tests rather than feature tests. RSpec is the [preferred testing suite] on GOV.UK and is favoured by many developers because the tests and testing output are easier to read.
+
+### Code structure
+The code structure will be standardised. As well as using the [default Rails directory structure], we propose the following standard for additional classes (e.g. services, api calls). The standard expands on the [conventions] documented in the developer docs.
+
+The host app will follow the default MVC structure of Rails applications. In addition to that:
+
+Under `/app`:
+
+models
+- Ingest data from yaml / JSON / CSVs rather than database.
```

Comment:
Suggestion:
  "Frontend apps do not have database records, so models here are created to store data received from api calls (eg content items from content store), process data that will be written to apis (eg feedback items to zendesk), and information ingested from YAML/JSON/CSVs. Unless it's not otherwise advisable, static data in the app will be stored in the locale files to promote localisation."

Reply:
The first bit sounds really good. Thanks. 👍 

What data are you think of here?
> static data in the app will be stored in the locale files 

I think static and dynamic content should be in the locale files, yes. But data? I think that should be in `config/data`. I don't think that needs to be mentioned in the models.